Mathlib/Tactic/Bound/Attribute.lean

```diff
@@ -0,0 +1,139 @@
+/-
+Copyright (c) 2024 Geoffrey Irving. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Geoffrey Irving
+-/
+import Mathlib.Algebra.Group.ZeroOne
+import Mathlib.Tactic.Bound.Init
+import Qq
+
+/-!
+# The `bound` attribute
+
+Any lemma tagged with `@[bound]` is registered as an apply rule for the `bound` tactic, by
+converting it to either `norm apply` or `safe apply <priority>`.  The classification is based
+on the number and types of the lemma's hypotheses.
+-/
+
+open Lean (MetaM)
+open Qq
+
+namespace Bound
+
+initialize Lean.registerTraceClass `bound.attribute
+
+variable {u : Lean.Level} {α : Q(Type u)}
+
+/-- Check if an expression is zero -/
+def isZero (e : Q($α)) : MetaM Bool :=
+  match e with
+  | ~q(@OfNat.ofNat.{u} _ (nat_lit 0) $i) => return true
+  | _ => return false
+
+/-- Map the arguments of an inequality expression to a score -/
+def ineqPriority (a b : Q($α)) : MetaM ℕ := do
+  return if (← isZero a) || (← isZero b) then 1 else 10
+
+/-- Map a hypothesis type to a score -/
+partial def hypPriority (hyp : Q(Prop)) : MetaM ℕ := do
+  match hyp with
+    -- Conjunctions add scores
+    | ~q($a ∧ $b) => pure <| (← hypPriority a) + (← hypPriority b)
+    -- Guessing (disjunction) gets a big penalty
+    | ~q($a ∨ $b) => pure <| 100 + (← hypPriority a) + (← hypPriority b)
```

Comment:
I just wanted to record it here for later. I'm not sure that a constant additive penalty is the appropriate solution here. It seems like you would want some sort of multiplicative penalty for anything that requires branching. I don't think this is enough of a problem to hold up the PR, but in case we have performance issues as a result of branching, I think this is a good place to look for the culprit in the future.

Reply:
That makes sense, but as a subtlety one would might to attach the multiplicative penalty to the Aesop rule that does the branching on the or generated by this rule, not this rule itself. Not sure how to represent that.